File: app/api/schemas/settlement.py
"""
Pydantic schemas for Settlement API endpoints.
"""
from pydantic import BaseModel

from app.api.schemas.period import MemberBalance


class SettlementBalanceResponse(BaseModel):
    """Schema for settlement balance response."""
    balances: list[MemberBalance]


class SettlementTransaction(BaseModel):
    """Schema for a settlement transaction."""
    date: str
    transaction_type: str
    amount: int  # Amount in cents (negative for refunds)
    description: str | None
    payer_id: int
    payer_name: str
    from_to: str  # "From" or "To"


class SettlementPlanResponse(BaseModel):
    """Schema for settlement plan response."""
    transactions: list[SettlementTransaction]

File: cli/lib/src/services/settlement.dart
// Service layer for settlement operations.

import 'package:built_collection/built_collection.dart';
import '../api/divvy_client.dart';
import '../ui/displays.dart';
import '../utils/formatting.dart';

/// Get settlement plan.
///
/// Parameters:
/// - [client]: The API client
/// - [periodId]: Optional period ID (defaults to current period)
///
/// Returns:
/// List of settlement transactions
Future<List<TransactionDisplay>> getSettlementPlan(
  DivvyClient client, {
  int? periodId,
}) async {
  final response = await client.settlement
      .getSettlementPlanApiV1SettlementPlanGet(periodId: periodId);
  final plan = response.data;
  if (plan == null) {
    return [];
  }

  return plan.transactions.map((tx) {
    final amountFormatted = formatAmountWithSign(tx.amount);
    return TransactionDisplay(
      date: tx.date,
      category: '',
      type: tx.transactionType,
      split: '',
      amount: tx.amount,
      amountFormatted: amountFormatted,
      description: tx.description ?? '',
      payer: tx.fromTo,
    );
  }).toList();
}

/// Get settlement balances.
///
/// Parameters:
/// - [client]: The API client
///
/// Returns:
/// List of member balances with descriptions
Future<List<({String memberName, String balanceDescription})>>
getSettlementBalances(DivvyClient client) async {
  final response = await client.settlement
      .getSettlementBalancesApiV1SettlementBalancesGet();
  final balances = response.data?.balances ?? BuiltList();
  return balances
      .map(
        (b) => (
          memberName: b.memberName,
          balanceDescription: b.balanceDescription,
        ),
      )
      .toList();
}
